src/task_impl/atomic_task.rs

```diff
@@ -7,7 +7,25 @@ use core::cell::UnsafeCell;
 use core::sync::atomic::AtomicUsize;
 use core::sync::atomic::Ordering::{Acquire, Release};
 
-/// A coordinated `Task` handle enabling concurrent operations on a task.
+/// A synchronization primitive for task notification.
+///
+/// `AtomicTask` will coordinate concurrent notifications with the consumer
+/// potentially "updating" the underlying task to notify. This is useful in
+/// scenarios where a computation completes in another thread and wants to
+/// notify the consumer, but the consumer is in the process of being migrated to
+/// a new logical task.
+///
+/// Roughly, consumers should call `park` before checking the result of a
```

Comment:
Yeah, that pattern was my original attempt. However, it doesn't really work in the face of the consumer **changing** its task in the same way that it does w/ `thread::park()`.

As you said, `thread::park()`  / `unpark()` behaves as a buffered notification. If `unpark` is called first, then that notification gets buffered until `park` is called to "consume" the notification.

AtomicTask scenario:

* Consumer calls `park` w/ task A.
* Consumer migrates to task B & checks computation result (not completed).
* Producer completes computation.
* Producer calls `notify`
* Consumer calls park` on task B.

The notification will be sent to task `A` and task `B` has deadlocked.

This primitives is more as a strategy to manage consumers migrating across tasks. This is why the consumer must call `park` before.

One disconnect might be that this PR leaves the original `park` naming. I'm open to naming suggestions.